bb8/src/api.rs

```diff
@@ -253,6 +268,38 @@ pub trait ManageConnection: Sized + Send + Sync + 'static {
     fn has_broken(&self, conn: &mut Self::Connection) -> bool;
 }
 
+/// A trait which provides functionality to initialize and cleanup a connection
+#[async_trait]
+pub trait CustomizeConnection<C: Send + 'static, E: 'static>:
+    std::fmt::Debug + Send + Sync + 'static
+{
+    /// Called with connections immediately after they are returned from
+    /// `ManageConnection::connect`.
+    ///
+    /// The default implementation simply returns `Ok(())`.
+    ///
+    /// # Errors
+    ///
+    /// If this method returns an error, the connection will be discarded.
+    #[allow(unused_variables)]
+    async fn on_acquire(&self, connection: &mut C) -> Result<(), E> {
+        Ok(())
+    }
+
+    /// Called with connections when they are removed from the pool.
+    ///
+    /// The connections may be broken (as reported by `ManageConnection::is_valid` or
+    /// `ManageConnection::has_broken`), or have simply timed out.
+    ///
+    /// The default implementation does nothing.
+    #[allow(unused_variables)]
+    fn on_release(&self, connection: &mut C) {}
```

Comment:
Yeah that makes sense. I gave that a shot and also made it return a `Result<(), E>` for consistency.

https://github.com/agersant/bb8/commit/5d5ff3210299d6838b75a11c5cbac0d5dcd0bca6

Unfortunately, this breaks two tests (annotated). I think none of this would be a problem if `put_back` was async, but it's called from `PooledConnection::drop` so that doesn't seem possible. Unsure where to go from there :(

Reply:
Hmm, that is annoying. Do you actually need `on_release()` for your use case? If not, I suggest we just make do with `on_acquire()` for this PR.

Reply:
Works for me!